#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [1]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [2]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train.csv")


## Data exploration

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [4]:
train.head()

laptop_ID Company            Product   TypeName  Inches  \
0       1104    Acer     Aspire ES1-523   Notebook    15.6   
1        114    Dell             XPS 13  Ultrabook    13.3   
2        172    Acer    Aspire A517-51G   Notebook    17.3   
3        918      HP      Elitebook 820    Netbook    12.5   
4        447  Lenovo  IdeaPad 320-15AST   Notebook    17.3   

                   ScreenResolution                         Cpu  Ram  \
0                          1366x768   AMD A8-Series 7410 2.2GHz  4GB   
1  Quad HD+ / Touchscreen 3200x1800  Intel Core i7 7560U 2.4GHz  8GB   
2       IPS Panel Full HD 1920x1080  Intel Core i5 8250U 1.6GHz  8GB   
3                 Full HD 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   
4                          1600x900   AMD A6-Series 9220 2.5GHz  8GB   

      Memory                           Gpu       OpSys  Weight  Price_euros  
0  500GB HDD                 AMD Radeon R5  Windows 10   2.4kg        387.0  
1  256GB SSD  Intel Iris Plus Graphics 640  Windows 10  1.23kg       1379.0  
2  256GB SSD          Nvidia GeForce MX150  Windows 10     3kg        854.0  
3  512GB SSD         Intel HD Graphics 620  Windows 10  1.26kg       1483.0  
4    1TB HDD                 AMD Radeon R4  Windows 10   2.8kg        519.0

In [11]:
train.tail()

laptop_ID Company                               Product  TypeName  \
907        996  Lenovo                     IdeaPad 320-15ABR  Notebook   
908        731    Dell                          Alienware 17    Gaming   
909        276    Asus                             Rog Strix    Gaming   
910        585     MSI                          GT73VR Titan    Gaming   
911       1173      HP  15-bw002nv (A6-9220/4GB/256GB/Radeon  Notebook   

     Inches       ScreenResolution                           Cpu   Ram  \
907    15.6      Full HD 1920x1080   AMD A12-Series 9720P 3.6GHz   6GB   
908    17.3  4K Ultra HD 3840x2160   Intel Core i7 7700HQ 2.8GHz  32GB   
909    17.3      Full HD 1920x1080           AMD Ryzen 1700 3GHz  16GB   
910    17.3      Full HD 1920x1080   Intel Core i7 7820HK 2.9GHz  16GB   
911    15.6      Full HD 1920x1080  AMD A6-Series A6-9220 2.5GHz   4GB   

                   Memory                      Gpu       OpSys  Weight  \
907             256GB SSD           AMD Radeon 530  Windows 10   2.2kg   
908    1TB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10  4.36kg   
909  256GB SSD +  1TB HDD        AMD Radeon RX 580  Windows 10   3.2kg   
910  512GB SSD +  1TB HDD  Nvidia GeForce GTX 1070  Windows 10  4.14kg   
911             256GB SSD           AMD Radeon 520  Windows 10  1.91kg   

     Price_euros  
907       549.00  
908      3659.40  
909      1549.00  
910      2729.00  
911       478.89

In [13]:
train.select_dtypes(include='number').corr()

laptop_ID    Inches  Price_euros
laptop_ID     1.000000 -0.076871     0.056520
Inches       -0.076871  1.000000     0.054903
Price_euros   0.056520  0.054903     1.000000

## Data processing

1. ScreenResolution (por ej: 1920x1080 es full HD)
2. CPU (tipo y GHz)
3. GPU (tipo y número, mientras mal alto mejor por lo que vi)
4. Memory (llevar todo a Gb, y separa entre HHD y SSD)

In [5]:
train.columns

Index(['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches',
       'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight',
       'Price_euros'],
      dtype='object')

In [10]:
train['Memory'].unique()

array(['500GB HDD', '256GB SSD', '512GB SSD', '1TB HDD', '1.0TB Hybrid',
       '128GB SSD +  1TB HDD', '256GB SSD +  1TB HDD', '128GB SSD',
       '256GB SSD +  500GB HDD', '256GB Flash Storage',
       '512GB SSD +  1TB HDD', '32GB Flash Storage', '64GB Flash Storage',
       '64GB SSD', '512GB SSD +  256GB SSD', '1TB SSD', '180GB SSD',
       '128GB SSD +  2TB HDD', '128GB Flash Storage', '2TB HDD',
       '1TB HDD +  1TB HDD', '32GB SSD', '16GB Flash Storage',
       '256GB SSD +  256GB SSD', '512GB SSD +  2TB HDD', '16GB SSD',
       '256GB SSD +  2TB HDD', '512GB Flash Storage', '508GB Hybrid',
       '64GB Flash Storage +  1TB HDD', '512GB SSD +  512GB SSD',
       '1TB SSD +  1TB HDD'], dtype=object)

In [8]:
train[train['OpSys'] == 'No OS']

laptop_ID Company                                Product   TypeName  \
6           63  Lenovo                     IdeaPad 520S-14IKB   Notebook   
14         272  Lenovo                     IdeaPad 320-15IKBN   Notebook   
15         196  Xiaomi                            Mi Notebook  Ultrabook   
40         176  Lenovo                      IdeaPad 320-15AST   Notebook   
67         601  Lenovo  V320-17ISK (i3-6006U/4GB/500GB/FHD/No   Notebook   
83         310  Lenovo                      IdeaPad 320-15IAP   Notebook   
108        554  Lenovo    V310-15ISK (i5-6200U/4GB/1TB/FHD/No   Notebook   
132         19  Lenovo                      IdeaPad 320-15IKB   Notebook   
133        792  Lenovo                      IdeaPad 110-15IBR   Notebook   
155        534  Lenovo                      IdeaPad 320-15ISK   Notebook   
188       1170  Lenovo        B51-80 (i5-6200U/8GB/1TB/Radeon   Notebook   
201        141  Lenovo                     IdeaPad 320-15IKBN   Notebook   
215       1219  Lenovo                      IdeaPad 310-15ISK   Notebook   
233        188  Xiaomi                            Mi Notebook   Notebook   
248        128      HP                                 250 G6   Notebook   
262        521  Xiaomi                            Mi Notebook   Notebook   
317       1235  Lenovo                      IdeaPad 320-17IKB   Notebook   
318       1191  Lenovo      V110-15ISK (i5-6200U/4GB/500GB/No   Notebook   
349        308  Lenovo                     IdeaPad 320-15IKBN   Notebook   
352        370  Lenovo                      IdeaPad 320-15ISK   Notebook   
389        139  Lenovo           V110-15IAP (N3350/4GB/1TB/No   Notebook   
400         78  Lenovo                     IdeaPad 320-15IKBN   Notebook   
429        874  Lenovo                     Legion Y520-15IKBN     Gaming   
437        733  Lenovo        V310-15IKB (i5-7200U/4GB/1TB/No   Notebook   
455        870  Lenovo                      IdeaPad 110-15ISK   Notebook   
495        634  Lenovo                     IdeaPad 320-15IKBN   Notebook   
499       1075    Asus                      Rog G752VL-GC088D     Gaming   
504        946      HP                                 250 G6   Notebook   
507        470  Lenovo                     IdeaPad 510s-14IKB   Notebook   
515        651    Asus                      ROG GL553VE-FY022     Gaming   
539        217      HP  15-bs018nq (i3-6006U/4GB/500GB/FHD/No   Notebook   
559        209  Lenovo                     Legion Y520-15IKBN     Gaming   
574        125  Lenovo                     IdeaPad 320-15IKBN   Notebook   
610        221  Lenovo                      IdeaPad 320-17IKB   Notebook   
650       1266  Lenovo  V510-15IKB (i5-7200U/8GB/256GB/FHD/No   Notebook   
707       1143  Lenovo                     Legion Y520-15IKBN     Gaming   
719         12      HP                                 250 G6   Notebook   
761         23      HP                                 255 G6   Notebook   
770       1182      HP                                 250 G4   Notebook   
775       1106      HP                                 250 G6   Notebook   
782        516  Lenovo                     Legion Y520-15IKBN     Gaming   
784        387  Lenovo  V310-15ISK (i3-6006U/4GB/128GB/FHD/No   Notebook   
792        362  Lenovo                      IdeaPad 320-15IAP   Notebook   
796        145  Lenovo                     Legion Y520-15IKBN     Gaming   
828        882  Lenovo                      IdeaPad 110-15ISK   Notebook   
881       1192  Lenovo                      IdeaPad 510-15IKB   Notebook   
895       1168  Lenovo        V110-15ISK (i3-6006U/4GB/1TB/No   Notebook   

     Inches             ScreenResolution  \
6      14.0  IPS Panel Full HD 1920x1080   
14     15.6            Full HD 1920x1080   
15     13.3  IPS Panel Full HD 1920x1080   
40     15.6                     1366x768   
67     17.3            Full HD 1920x1080   
83     15.6                     1366x768   
108    15.6            Full HD 1920x1080   
132 

TypeError: DataFrame.corr() got an unexpected keyword argument 'include'

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [11]:
train = train[['laptop_ID','Inches','Price_euros']]

In [12]:
train

laptop_ID  Inches  Price_euros
0         1104    15.6       387.00
1          114    13.3      1379.00
2          172    17.3       854.00
3          918    12.5      1483.00
4          447    17.3       519.00
..         ...     ...          ...
907        996    15.6       549.00
908        731    17.3      3659.40
909        276    17.3      1549.00
910        585    17.3      2729.00
911       1173    15.6       478.89

[912 rows x 3 columns]

In [13]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 2)

In [14]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [17]:
X_train

laptop_ID  Inches
25          26    15.6
84         776    15.6
10        1263    15.6
342        182    15.6
890        321    13.3
..         ...     ...
106        394    15.6
270        441    15.6
860        955    17.3
435        210    15.6
102        673    15.6

[729 rows x 2 columns]

### 3. Asignar el modelo (vacío) a una variable


In [22]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

In [23]:
model.fit(X,y)

LinearRegression()

In [20]:
pred = model.predict(X_test)
pred

array([1120.43868276, 1069.79534654, 1070.20842432, 1160.62136315,
       1208.45577045, 1022.56227854, 1154.97198544, 1077.94506983,
       1074.00873993, 1084.55546588, 1135.07084845, 1128.48391161,
       1102.85078136, 1054.32277823, 1131.84884174, 1137.90208507,
       1085.50151446, 1055.44096704, 1172.81694567, 1116.91773368,
       1150.18028315, 1160.67245885, 1083.34429783, 1091.0275446 ,
       1124.35313378, 1154.75565863, 1101.65688611, 1126.25329158,
       1072.19119768, 1119.5614315 , 1060.10126111, 1030.7727385 ,
       1176.36941461, 1144.34609849, 1117.30850377, 1081.93983336,
       1168.60355229, 1108.67920842, 1107.55986809, 1173.31263901,
       1169.59493897, 1106.80711591, 1139.93595413, 1140.711014  ,
       1173.89094791, 1060.22115413, 1083.09645116, 1136.38348519,
       1091.60585349, 1098.54556025, 1103.14857221, 1071.7047165 ,
       1089.12738679, 1130.05360719, 1137.13623737, 1103.42909025,
       1121.40200414, 1081.19629335, 1066.41732088, 1186.72787

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [21]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE :',np.sqrt(mean_squared_error(y_test,pred)))

RMSE : 705.2026860694585


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [35]:
X_pred = pd.read_csv("data/test.csv")
X_pred.head()

laptop_ID   Company          Product            TypeName  Inches   
0        750   Toshiba  Tecra X40-D-10G            Notebook    14.0  \
1        726  Mediacom    Smartbook 142            Notebook    14.0   
2        633  Mediacom    SmartBook 140            Notebook    14.0   
3        363      Dell    Inspiron 7577              Gaming    15.6   
4        319      Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                            ScreenResolution   
0  IPS Panel Full HD / Touchscreen 1920x1080  \
1                IPS Panel Full HD 1920x1200   
2                          Full HD 1920x1080   
3                          Full HD 1920x1080   
4                       Touchscreen 1366x768   

                                    Cpu  Ram              Memory   
0            Intel Core i5 7200U 2.5GHz  8GB           256GB SSD  \
1              Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
2           Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
3           Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
4  Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                       Gpu       OpSys  Weight  
0    Intel HD Graphics 620  Windows 10  1.25kg  
1    Intel HD Graphics 400  Windows 10   1.4kg  
2        Intel HD Graphics  Windows 10   1.4kg  
3  Nvidia GeForce GTX 1050  Windows 10  2.65kg  
4    Intel HD Graphics 500  Windows 10   1.1kg

In [36]:
X_pred=X_pred[['laptop_ID','Inches']]

In [37]:
X_pred

laptop_ID  Inches
0          750    14.0
1          726    14.0
2          633    14.0
3          363    15.6
4          319    11.6
..         ...     ...
386        693    15.6
387       1132    12.5
388         49    15.6
389       1243    15.6
390        924    12.5

[391 rows x 2 columns]

**¿Por qué da error?** 

In [38]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1097.6423004 , 1095.00755928, 1084.79793741, 1101.01825451,
        981.53500887, 1062.70472731, 1123.17875843, 1136.91660235,
       1100.68891187, 1171.05845611, 1061.49713762, 1039.67799563,
       1120.5592512 , 1084.55112248, 1067.42530516, 1144.60126397,
       1008.086748  , 1087.21311678, 1117.59516743, 1166.6672209 ,
       1136.28517025, 1137.01114934, 1201.1384173 , 1195.5395924 ,
       1155.57935199, 1095.22712104, 1205.96877603, 1141.07304191,
       1087.734768  , 1145.58929189, 1069.84048452, 1117.48538655,
       1169.08240026, 1183.02457206, 1166.99656354, 1048.59589281,
       1143.28389341, 1069.57766753, 1078.759989  , 1072.74768646,
       1196.62216731, 1063.58297435, 1030.59182845,  995.79128941,
       1065.55903019, 1129.78084514, 1071.26763597, 1167.64001493,
       1224.39673002, 1248.32896191, 1181.26807797, 1092.56512673,
       1090.69885177, 1189.17230135, 1066.32749635, 1158.8727784 ,
       1102.11606331, 1053.33170455, 1106.66997846, 1072.30856

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [43]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1097.6423004 , 1095.00755928, 1084.79793741, 1101.01825451,
        981.53500887, 1062.70472731, 1123.17875843, 1136.91660235,
       1100.68891187, 1171.05845611, 1061.49713762, 1039.67799563,
       1120.5592512 , 1084.55112248, 1067.42530516, 1144.60126397,
       1008.086748  , 1087.21311678, 1117.59516743, 1166.6672209 ,
       1136.28517025, 1137.01114934, 1201.1384173 , 1195.5395924 ,
       1155.57935199, 1095.22712104, 1205.96877603, 1141.07304191,
       1087.734768  , 1145.58929189, 1069.84048452, 1117.48538655,
       1169.08240026, 1183.02457206, 1166.99656354, 1048.59589281,
       1143.28389341, 1069.57766753, 1078.759989  , 1072.74768646,
       1196.62216731, 1063.58297435, 1030.59182845,  995.79128941,
       1065.55903019, 1129.78084514, 1071.26763597, 1167.64001493,
       1224.39673002, 1248.32896191, 1181.26807797, 1092.56512673,
       1090.69885177, 1189.17230135, 1066.32749635, 1158.8727784 ,
       1102.11606331, 1053.33170455, 1106.66997846, 1072.30856

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [44]:
sample = pd.read_csv("data/sample_submission.csv")

In [45]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [46]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [53]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [55]:
submission.head()

laptop_ID  Price_euros
0        750  1097.642300
1        726  1095.007559
2        633  1084.797937
3        363  1101.018255
4        319   981.535009

In [28]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [56]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [57]:
chequeator(submission)

You're ready to submit!
